utils/lib/rr_graph/graph.py

```diff
@@ -1,4 +1,44 @@
 #!/usr/bin/env python3
+'''
+rr_graph docs: http://docs.verilogtorouting.org/en/latest/vpr/file_formats/
+etree docs: http://lxml.de/api/lxml.etree-module.html
+
+general philosiphy
+Data structures should generally be conceptual objects rather than direct XML manipulation
+Generally one class per XML node type
+However, GraphIdsMap is special cased, operating on XML directly and holding
+
+Class list
+class Pin(MostlyReadOnly):
+    xml: can import but doesn't keep track of a node
+class PinClass(MostlyReadOnly):
+    xml: can import but doesn't keep track of a node
+class BlockType(MostlyReadOnly):
+    xml: can import but doesn't keep track of a node
+class Block(MostlyReadOnly):
+    for <grid_loc>
+class BlockGrid:
+    Was: BlockGraph
+    xml: nothing, handled by intneral Block objects though
+class GraphIdsMap:
+    holds pins + edges
+    xml: updated as pins are added
+        inconsistent with the rest of the project
```

Comment:
FYI -- The reason for the inconsistency is that the other parts are not designed to allow modification/changing the XML.